Add tests derived from RFC 3720, section B.4.

tseaver opened the pull request.

tseaver committed:
Add tests derived from RFC 3720, section B.4.

dhermes commented on tests/test___init__.py:
```diff
@@ -0,0 +1,38 @@
+import pytest
```
License header needed

tseaver replied:
e8bbd2b

dhermes commented on tests/test___init__.py:
```diff
@@ -0,0 +1,38 @@
+
+
+def _call_extend(crc, chunk):
+    import crc32c
```
May as well just import at import time?

tseaver replied:
e8bbd2b

dhermes commented on tests/test___init__.py:
```diff
@@ -0,0 +1,38 @@
+
+
+# From: https://tools.ietf.org/html/rfc3720#appendix-B.4
+iscsi_scsi_read_10_command_pdu = [
```
PEP8, all caps var name?

tseaver replied:
e8bbd2b

dhermes reviewed (commented):
1. Can you add a test that breaks `iscsi_scsi_read_10_command_pdu` into a few chunks and calls `extend()` multiple times, then compares against `value()` output?
1. Could you add a test to see how much `cffi` freaks out if the wrong types are passed in?

dhermes commented:
@tseaver Is there a plan to do more re-factoring once these tests are in place?

tseaver commented:
> Is there a plan to do more re-factoring once these tests are in place?

Yes, I plan to make the args explict for `value` and `extend`, and add in the hashlib-like wrapper.

dhermes commented:
@tseaver I just noticed that AppVeyor decided to run for this (because it needed to build wheels).

Care to add a test run in AppVeyor? It should be easy since each environment already builds the wheels.

```
%PYTHON3X%/python \
  -m pip \
  install \
  --no-index \
  --find-links . \
  py_crc32c
```

tseaver committed:
Ignore 'usr/' directory.

Used as an install target for the 'libcrc32c.so' shared library,
and also maybe as a virtualenv.

tseaver committed:
Try to get the build more reproducible.

Document the steps for installing in development for testing.

tseaver committed:
Review comments:

- Add license header.
- PEP 8 for constant name.
- Import 'crc32c' at module scope.
- Add a test for multiple 'extend' calls.

tseaver commented:
> Can you add a test that breaks iscsi_scsi_read_10_command_pdu into a few chunks and calls extend() multiple times, then compares against value() output?

Added in e8bbd2b

dhermes commented on README.md:
```diff
@@ -100,3 +100,28 @@ On Windows: TBD.
+>   ..
+$ make all install
+$ cd ../..
+$ python3.Y -m venv usr
```
1. Is `venv` preferred over `virtualenv`? (I was under the impression that `venv` was not well-liked)
1. Can you name it something other than `usr`, like `venv`?

tseaver replied:
> 1. Is venv preferred over virtualenv? (I was under the impression that venv was not well-liked)

`venv` is present only on Python >= 3.3, but it is ubiquitous there (don't have to bootstrap it).

> 2. Can you name it something other than usr, like venv?

OK.  I figured reusing `usr` made sense (share the `bin` and `lib` directories already there after the `make` step above.

dhermes commented on src/crc32c_build.py:
```diff
@@ -23,9 +23,11 @@ def get_kwargs():
     return {
         "library_dirs": library_dirs,
         "include_dirs": [os.path.join(install_prefix, "include")],
+        "extra_link_args": ["-Wl,-rpath={}".format(rpath)],
```
We should verify that this doesn't "break" the `manylinux` wheels? I can do the same for OS X?

tseaver commented:
> Could you add a test to see how much cffi freaks out if the wrong types are passed in?

I don't know how useful a test would be for asserting this traceback:

```python
>>> import crc32c
>>> crc32c.value('', 0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/tseaver/projects/agendaless/Google/src/py-crc32c/src/crc32c/__init__.py", line 24, in value
    return crc32c._crc32c_cffi.lib.crc32c_value(*args, **kwargs)
TypeError: initializer for ctype 'uint8_t *' must be a bytes or list or tuple, not str
```

dhermes commented on tests/test___init__.py:
```diff
@@ -1,38 +1,55 @@
-    import crc32c
-    return crc32c.value(chunk, len(chunk))
+def test_extend_w_empty_chunk():
+    crc = 0
```
Unused variable here

dhermes reviewed (approved):
LGTM other than a few nits

dhermes commented:
```
TypeError: initializer for ctype 'uint8_t *' must be a bytes or list or tuple, not str
```

is a very helpful error message though!

tseaver committed:
Run tests on Appveyor.

tseaver committed:
Test list-of-integers, too.

tseaver committed:
Review comments: [ci skip]

- Use 'virtualenv' for portability.
- Keep the Python virtual environment separate from the C install target.
- Fix unused variable in testcase.

dhermes commented on README.md:
```diff
@@ -100,3 +100,44 @@ On Windows: TBD.
+$ venv/bin/py.test tests/
+============================= test session starts ==============================
+platform linux -- Python 3.6.7, pytest-3.10.0, py-1.7.0, pluggy-0.8.0
+rootdir: /home/tseaver/projects/agendaless/Google/src/py-crc32c, inifile:
```
@tseaver Path snuck in here

dhermes replied:
@tseaver Can you remove this in #4?

tseaver replied:
43207e0
